arch/lkl/kernel/cpu.c

```diff
@@ -128,9 +132,11 @@ void lkl_cpu_put(void)
 	}
 
 	if (test_ti_thread_flag(current_thread_info(), TIF_HOST_THREAD) &&
-	    !single_task_running() && cpu.count == 1) {
-		if (in_interrupt())
+		!single_task_running() && cpu.count == 1) {
```

Comment:
This code is actually unchanged. I added a call to print the lkl_cpu object state if it does fail as that helps debugging.
I believe all this stuff with locking the cpu has wider application than signals. Also since SEGV and FPE are signals a thread can send itself a signal, and I think that the various APIs which expose a thread group (eg tgkill) can send it to the only thread.